src/java/org/apache/cassandra/service/accord/txn/AbstractKeySorted.java

```diff
@@ -55,16 +54,39 @@ public AbstractKeySorted(List<T> items)
         items.toArray(arr);
         Arrays.sort(arr, this::compare);
         this.items = arr;
-        validateOrder();
-        this.itemKeys = extractItemKeys();
+        int duplicates = validateOrder();
+        this.itemKeys = extractItemKeys(duplicates);
     }
 
-    private Keys extractItemKeys()
+    private Keys extractItemKeys(int duplicates)
     {
-        Key[] keys = new Key[size()];
-        for (int i = 0 ; i < keys.length ; ++i)
-            keys[i] = getKey(items[i]);
-        return Keys.ofSorted(keys);
+        PartitionKey[] partitionKeys = new PartitionKey[items.length - duplicates];
```

Comment:
given the duplicate case is likely to be uncommon, we could simplify this and just call `Keys.ofSorted`? I don't have a super strong opinion though, this is totally fine, but it is a shame to have dedicated de-dup code. We could also introduce a version of `Keys.ofSorted()` that accepts a size, the getter function and the input array to transform?

Reply:
Alternatively, we can avoid additional key comparisons by using `ArrayBuffers.cachedKeys` to accumulate the keys. I think it would still be nice to offer this as a static builder in either `Keys` or `SortedArrays` (or in both, with the former using the latter to build the array) so that it can be reused. WDYT?

Reply:
I was just trying to reuse the work done to ensure that is sorted. The simpler approach is to just remove the sorting check and rely on the one in `Keys.ofSorted`